### PR TITLE
Type verification in tests

### PR DESCRIPTION
**Pull request by davidwrighton**
- Enable type layout verification in crossgen2 test passes
- Fix issues noted from last attempt to do so
- Do not generate type layout checks or verification if the offset is 24 bits or larger
- Only load approx enclosing type when verifying, instead of enclosing type (this avoids asserts)
- HFA encoding was not correct for Arm64, and did not have R2R defined constants

Do not merge until this has passed manual checks against all of the various crossgen2 test pipelines

- [x] Crossgen2 pipeline tested
- [x] Crossgen2 Composite pipeline tested

### CODE REVIEW

**Comment by AntonLapounov** [comment]
Why do we need to introduce `ReadyToRunHFAElemType` if it is the same as `CorInfoHFAElemType`?

File: src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/FieldFixupSignature.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ namespace ILCompiler.DependencyAnalysis.ReadyToRun
 {
     public class FieldFixupSignature : Signature
     {
+        public const int MaxCheckableOffset = 0x1FFFFFFF;
```
Comment (AntonLapounov) [inline src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/FieldFixupSignature.cs:17]: What was the issue with larger offsets?  Are they now unchecked?

File: src/coreclr/src/tools/Common/Internal/Runtime/ReadyToRunConstants.cs [hunk 2]
```diff
@@ -314,6 +314,17 @@ public enum ReadyToRunHelper
         TypeHandleToRuntimeTypeHandle,
     }
 
+    // Enum used for HFA type recognition.
+    // Supported across architectures, so that it can be used in altjits and cross-compilation.
+    public enum ReadyToRunHFAElemType
+    {
+        None,
```
Comment (AntonLapounov) [inline src/coreclr/src/tools/Common/Internal/Runtime/ReadyToRunConstants.cs:321]: Nit: I would use `None = 0` for consistency with the following lines and the `ReadyToRunHFAElemType` enumeration.

File: src/coreclr/src/inc/readytorun.h [hunk 3]
```diff
@@ -395,4 +395,13 @@ enum ReadyToRunRuntimeConstants : DWORD
     READYTORUN_ReversePInvokeTransitionFrameSizeInPointerUnits = 2
 };
 
+enum ReadyToRunHFAElemType : DWORD
+{
+    READYTORUN_HFA_ELEMTYPE_None = 0,
+    READYTORUN_HFA_ELEMTYPE_Float = 1,
+    READYTORUN_HFA_ELEMTYPE_Double = 2,
```
Comment (AntonLapounov) [inline src/coreclr/src/inc/readytorun.h:402]: Note that `float`/`double` naming is specific to C# (and C/C++).  For instance, in F# `float` is a [synonym](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/basic-types) for `double`.  VB .NET has the `Double` keyword, but no `Float` one. For that reason I would prefer either `Float32`/`Float64` as in the CIL assembler or `Single`/`Double` as in the corresponding .NET type names.

**Review by trylek — APPROVED** [review]
LGTM modulo Anton's feedback

File: src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/FieldFixupSignature.cs (hunk 1 above)
Comment (davidwrighton) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/FieldFixupSignature.cs:17]: They are now unchecked. The problem is that the encoding uses the compressed uint encoding from metadata which cannot encode a number greater than 0x1FFFFFFF.

In practice, the need to do this is so rare, that it isn't worth changing the encoding.

**Comment by davidwrighton** [comment]
> Why do we need to introduce `ReadyToRunHFAElemType` if it is the same as `CorInfoHFAElemType`?

All contents of the ReadyToRun file format should be either ECMA 335 constants, or defined as part of the ready to run headers.

**Review by jkoritzinsky — APPROVED** [review]
fieldmarshaler.cpp changes look good to me!

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky I see some validation here for crossgen2, what about the runtime itself? I would assume the runtime needs a new test or to update one.

**Comment by jkoritzinsky** [comment]
I don't believe we have a type layout test in the runtime for this case.

The crossgen2 test will catch any unintentional layout changes in this area since crossgen2 has a parallel implementation, so a runtime test isn't as immediately important. It would still be worth adding though I think.

File: src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/FieldFixupSignature.cs (hunk 1 above)
Comment (AntonLapounov) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/FieldFixupSignature.cs:17]: Shouldn't we fail compilation for large offsets then?  I see throwing a `RequiresRuntimeJitException` in one `if` branch, but not in others.

**Pull request merged**